internal/credentials/docker.go

```diff
@@ -1,13 +1,56 @@
 package credentials
 
-type DockerConfig map[string]DockerConfigEntry
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"strings"
 
-type DockerConfigEntry struct {
-	Username string `json:"username"`
-	Password string `json:"password"`
-}
+	"github.com/docker/docker/api/types"
+	"github.com/pkg/errors"
+)
+
+// AuthConfigs is a map of registry urls to authentication credentials.
+type AuthConfigs map[string]types.AuthConfig
 
-// DockerConfigJSON models the structure of .dockerconfigfile data
+// DockerConfigJSON models the structure of .dockerconfigfile data.
 type DockerConfigJSON struct {
-	Auths DockerConfig `json:"auths"`
+	Auths AuthConfigs `json:"auths"`
+}
+
+// ExtractDockerAuth will return a username/password combination from a JSON-representation of a Docker config file.
+// This function will process credentials from any of the possible auth-sources within that configuration.
+func ExtractDockerAuth(input []byte, host string) (string, string, error) {
```

Comment:
Can we use `docker/cli` for this or does the API not quite fit?

Reply:
I was thinking something like (haven't tested it)
```
auth := configfile.New().LoadFromReader(input).GetAuthConfig(host)
return auth.Username, auth.Password, ...
```
It would basically handle the decode / `Auth` stuff for you, not the credential helper stuff.